docs/hot_module_reloading.md

```diff
@@ -0,0 +1,190 @@
+js modules on the page for changes to take effect.
+
+If something depends on changed module A, it may have some instances of A's types, functions and
+objects closured, so reloading A itself doesnt invalidate closure. In basic configuration
+hot-reloading reloads all modules in chain from changed module up to the root module of applications
+and reruns `main()` function for changes to take effect. This means that without special measures
+state of your won't be preserved, nevertheless hot-reloading is much faster than full page
+reloading.
+     
+## Turning hot-reloading on
+`build_runner` server has built-in support of hot reloading. To activate it just run `serve` command
+with `--hot-reload` option.
+
+## Using hooks to handle module reloading
+
+Each module where changes are detected is marked as invalidated. For each invalidated module three
+hooks will be called to determinate, is it possible to handle it's reloading, or all parents (other
+modules that depends on this one) should be marked as invalidated too. To implement the hook all you
+need to do is to define a top-level publicly exported function with name and signature of a hook.
+This hooks are:
+
+### `Object hot$onDestroy();`
+
+This function will be called on old version of module before unloading.
+
+Implement this function with any code to release resources before destroy.
+
+Any object returned from this function will be passed to update hooks. Use
+it to save any state you need to be preserved between hot reloadings.
+Try do not use any custom types here, as it might prevent their code from
+reloading. Better serialise to JSON or plain types.
+
+### `bool hot$onSelfUpdate([Object data])`
+
+This function will be called on new version of module after reloading.
+
+If any state was saved from previous version, it will be passed to `data`.
+
+Implement this function to handle update of the module itself.
+
+May return nullable bool. To indicate that reload completes successfully
+return `true`. To indicate that hot-reload is undoable return `false` - this
+will lead to full page reload. If `null` returned, reloading will be
+propagated to parent.
+
+### `bool hot$onChildUpdate(String childPath, Object child, [Object data]);`
+
+This function will be called on current version of module parent after child
+reloading.
+
+The name of the child library will be provided in `childPath`. New version of child
+module object will be provided in `child`.
+If any state was saved from previous version, it will be passed to `data`.
+
+Implement this function to handle update of child modules.
+
+May return nullable bool. To indicate that reload of child completes
+successfully return `true`. To indicate that hot-reload is undoable for this
+child return `false` - this will lead to full page reload. If `null` returned,
+reloading will be propagated to current module itself.
+
+Accessing properties of provided `child` object is tricky thing. As dart libraries have no type
```

Comment:
nit: swap the ordering of this paragraph and the one above (this one is more important imo)